# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

In [2]:
path="G:\My Drive\Jun\Mechine Learning\Data Science\Guided Capstone\Case study - skiing resort"
os.chdir(path) 

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [3]:
df = pd.read_csv("step3_output.csv")
print(df.head(5))

   Unnamed: 0                 Name    state  summit_elev  vertical_drop  \
0           0       Alyeska Resort   Alaska         3939           2500   
1           1  Eaglecrest Ski Area   Alaska         2600           1540   
2           2     Hilltop Ski Area   Alaska         2090            294   
3           3     Arizona Snowbowl  Arizona        11500           2300   
4           4  Sunrise Park Resort  Arizona        11100           1800   

   trams  fastEight  fastSixes  fastQuads  quad  ...  SkiableTerrain_ac  \
0      1        0.0          0          2     2  ...             1610.0   
1      0        0.0          0          0     0  ...              640.0   
2      0        0.0          0          0     0  ...               30.0   
3      0        0.0          1          0     2  ...              777.0   
4      0        0.0          0          1     2  ...              800.0   

   Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  \
0           113.0

In [34]:
df.drop('Unnamed: 0',axis=1)

Name    state  summit_elev  vertical_drop  \
0                       Alyeska Resort   Alaska         3939           2500   
1                  Eaglecrest Ski Area   Alaska         2600           1540   
2                     Hilltop Ski Area   Alaska         2090            294   
3                     Arizona Snowbowl  Arizona        11500           2300   
4                  Sunrise Park Resort  Arizona        11100           1800   
..                                 ...      ...          ...            ...   
325               Meadowlark Ski Lodge  Wyoming         9500           1000   
326          Sleeping Giant Ski Resort  Wyoming         7428            810   
327                   Snow King Resort  Wyoming         7808           1571   
328  Snowy Range Ski & Recreation Area  Wyoming         9663            990   
329                White Pine Ski Area  Wyoming         9500           1100   

     trams  fastEight  fastSixes  fastQuads  quad  triple  ...  \
0        1        0.0          0          2     2       0  ...   
1        0        0.0          0          0     0       0  ...   
2        0        0.0          0          0     0       1  ...   
3        0        0.0          1          0     2       2  ...   
4        0        0.0          0          1     2       3  ...   
..     ...        ...        ...        ...   ...     ...  ...   
325      0        0.0          0          0     0       1  ...   
326      0        0.0          0          0     0       1  ...   
327      0        0.0          0          0     1       1  ...   
328      0        0.0          0          0     0       1  ...   
329      0        0.0          0          0     0       2  ...   

     SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0               1610.0      113.000000        150.000000       60.0   
1                640.0       60.000000         45.000000       44.0   
2                 30.0       30.000000        150.000000       36.0   
3                777.0      104.000000        122.000000       81.0   
4                800.0       80.000000        115.000000       49.0   
..                 ...             ...               ...        ...   
325              300.0      174.873239        115.103943        9.0   
326              184.0       18.000000         61.000000       81.0   
327              400.0      250.000000        121.000000       80.0   
328               75.0       30.000000        131.000000       59.0   
329              370.0      174.873239        115.103943       81.0   

     averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0         669.000000     65.000000      85.00000         150.000000   
1         350.000000     47.000000      53.00000          90.000000   
2          69.000000     30.000000      34.00000         152.000000   
3         260.000000     89.000000      89.00000         122.000000   
4         250.000000     74.000000      78.00000         104.000000   
..               ...           ...           ...                ...   
325       185.316456     57.916957      64.16681         120.053004   
326       310.000000     42.000000      42.00000          77.000000   
327       300.000000     59.000000      59.00000         123.000000   
328       250.000000     49.000000      49.00000         120.053004   
329       150.000000     57.916957      49.00000         120.053004   

     NightSkiing_ac  clusters  
0             550.0         1  
1               0.0         1  
2              30.0         1  
3               0.0         0  
4              80.0         0  
..              ...       ...  
325             0.0         0  
326             0.0         0  
327           110.0         0  
328             0.0         0  
329             0.0         0  

[330 rows x 26 columns]

In [35]:
df.describe().T

count         mean          std    min      25%  \
Unnamed: 0         330.0   164.500000    95.407023    0.0    82.25   
summit_elev        330.0  4591.818182  3735.535934  315.0  1403.75   
vertical_drop      330.0  1215.427273   947.864557   60.0   461.25   
trams              330.0     0.172727     0.559946    0.0     0.00   
fastEight          330.0     0.003030     0.055048    0.0     0.00   
fastSixes          330.0     0.184848     0.651685    0.0     0.00   
fastQuads          330.0     1.018182     2.198294    0.0     0.00   
quad               330.0     0.933333     1.312245    0.0     0.00   
triple             330.0     1.500000     1.619130    0.0     0.00   
double             330.0     1.833333     1.815028    0.0     1.00   
surface            330.0     2.621212     2.059636    0.0     1.00   
total_chairs       330.0     8.266667     5.798683    0.0     5.00   
Runs               330.0    48.214724    46.081366    3.0    19.00   
TerrainParks       330.0     2.820789     1.845918    1.0     2.00   
LongestRun_mi      330.0     1.433231     1.147352    0.0     0.60   
SkiableTerrain_ac  330.0   739.801223  1807.868079    8.0    85.00   
Snow Making_ac     330.0   174.873239   242.378848    2.0    55.00   
daysOpenLastYear   330.0   115.103943    32.231212    3.0   100.00   
yearsOpen          330.0    63.656535   109.263494    6.0    50.00   
averageSnowfall    330.0   185.316456   133.424094   18.0    71.25   
AdultWeekday       330.0    57.916957    23.898798   15.0    43.00   
AdultWeekend       330.0    64.166810    22.571324   17.0    49.00   
projectedDaysOpen  330.0   120.053004    28.742976   30.0   101.25   
NightSkiing_ac     330.0    56.890909    93.464882    0.0     0.00   
clusters           330.0     0.651515     0.483541    0.0     0.00   

                           50%          75%      max  
Unnamed: 0          164.500000   246.750000    329.0  
summit_elev        3127.500000  7806.000000  13487.0  
vertical_drop       964.500000  1800.000000   4425.0  
trams                 0.000000     0.000000      4.0  
fastEight             0.000000     0.000000      1.0  
fastSixes             0.000000     0.000000      6.0  
fastQuads             0.000000     1.000000     15.0  
quad                  0.000000     1.000000      8.0  
triple                1.000000     2.000000      8.0  
double                1.000000     3.000000     14.0  
surface               2.000000     3.000000     15.0  
total_chairs          7.000000    10.000000     41.0  
Runs                 34.000000    60.000000    341.0  
TerrainParks          2.820789     3.000000     14.0  
LongestRun_mi         1.000000     2.000000      6.0  
SkiableTerrain_ac   200.000000   739.801223  26819.0  
Snow Making_ac      132.500000   192.000000   3379.0  
daysOpenLastYear    115.103943   130.000000    305.0  
yearsOpen            58.000000    68.750000   2019.0  
averageSnowfall     158.000000   278.750000    669.0  
AdultWeekday         57.916957    67.000000    179.0  
AdultWeekend         64.166810    72.000000    179.0  
projectedDaysOpen   120.053004   136.000000    305.0  
NightSkiing_ac       27.000000    83.750000    650.0  
clusters              1.000000     1.000000      2.0

In [36]:
print(df['state'])

0       Alaska
1       Alaska
2       Alaska
3      Arizona
4      Arizona
        ...   
325    Wyoming
326    Wyoming
327    Wyoming
328    Wyoming
329    Wyoming
Name: state, Length: 330, dtype: object


## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [37]:
df1 = pd.concat([pd.get_dummies(df['state']), df.drop('state', axis=1)], axis=1)

In [38]:
df1.shape
df1.describe().T

count         mean          std    min      25%  \
Alaska             330.0     0.009091     0.095056    0.0     0.00   
Arizona            330.0     0.006061     0.077731    0.0     0.00   
California         330.0     0.063636     0.244475    0.0     0.00   
Colorado           330.0     0.066667     0.249823    0.0     0.00   
Connecticut        330.0     0.015152     0.122341    0.0     0.00   
Idaho              330.0     0.036364     0.187478    0.0     0.00   
Illinois           330.0     0.012121     0.109593    0.0     0.00   
Indiana            330.0     0.006061     0.077731    0.0     0.00   
Iowa               330.0     0.009091     0.095056    0.0     0.00   
Maine              330.0     0.027273     0.163124    0.0     0.00   
Maryland           330.0     0.003030     0.055048    0.0     0.00   
Massachusetts      330.0     0.033333     0.179778    0.0     0.00   
Michigan           330.0     0.087879     0.283548    0.0     0.00   
Minnesota          330.0     0.042424     0.201861    0.0     0.00   
Missouri           330.0     0.006061     0.077731    0.0     0.00   
Montana            330.0     0.036364     0.187478    0.0     0.00   
Nevada             330.0     0.012121     0.109593    0.0     0.00   
New Hampshire      330.0     0.048485     0.215115    0.0     0.00   
New Jersey         330.0     0.006061     0.077731    0.0     0.00   
New Mexico         330.0     0.027273     0.163124    0.0     0.00   
New York           330.0     0.100000     0.300456    0.0     0.00   
North Carolina     330.0     0.018182     0.133811    0.0     0.00   
Ohio               330.0     0.015152     0.122341    0.0     0.00   
Oregon             330.0     0.030303     0.171680    0.0     0.00   
Pennsylvania       330.0     0.057576     0.233293    0.0     0.00   
Rhode Island       330.0     0.003030     0.055048    0.0     0.00   
South Dakota       330.0     0.006061     0.077731    0.0     0.00   
Tennessee          330.0     0.003030     0.055048    0.0     0.00   
Utah               330.0     0.039394     0.194826    0.0     0.00   
Vermont            330.0     0.045455     0.208615    0.0     0.00   
Virginia           330.0     0.012121     0.109593    0.0     0.00   
Washington         330.0     0.030303     0.171680    0.0     0.00   
West Virginia      330.0     0.012121     0.109593    0.0     0.00   
Wisconsin          330.0     0.048485     0.215115    0.0     0.00   
Wyoming            330.0     0.024242     0.154035    0.0     0.00   
Unnamed: 0         330.0   164.500000    95.407023    0.0    82.25   
summit_elev        330.0  4591.818182  3735.535934  315.0  1403.75   
vertical_drop      330.0  1215.427273   947.864557   60.0   461.25   
trams              330.0     0.172727     0.559946    0.0     0.00   
fastEight          330.0     0.003030     0.055048    0.0     0.00   
fastSixes          330.0     0.184848     0.651685    0.0     0.00   
fastQuads          330.0     1.018182     2.198294    0.0     0.00   
quad               330.0     0.933333     1.312245    0.0     0.00   
triple             330.0     1.500000     1.619130    0.0     0.00   
double             330.0     1.833333     1.815028    0.0     1.00   
surface            330.0     2.621212     2.059636    0.0     1.00   
total_chairs       330.0     8.266667     5.798683    0.0     5.00   
Runs               330.0    48.214724    46.081366    3.0    19.00   
TerrainParks       330.0     2.820789     1.845918    1.0     2.00   
LongestRun_mi      330.0     1.433231     1.147352    0.0     0.60   
SkiableTerrain_ac  330.0   739.801223  1807.868079    8.0    85.00   
Snow Making_ac     330.0   174.873239   242.378848    2.0    55.00   
daysOpenLastYear   330.0   115.103943    32.231212    3.0   100.00   
yearsOpen          330.0    63.656535   109.263494    6.0    50.00   
averageSnowfall    330.0   185.316456   133.424094   18.0    71.25   
AdultWeekday       330.0    57.916957    23.898798   15.0    43.00   
AdultWeekend       330.0    6

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [39]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df1.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df1.AdultWeekend 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [40]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [41]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [42]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [43]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

In [44]:
from sklearn.metrics import explained_variance_score, mean_absolute_error

In [45]:
print(explained_variance_score(y_test, y_pred))

0.9438633227455244


In [46]:
print(mean_absolute_error(y_test, y_pred))

4.725990783507439


**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [47]:
print(lm.intercept_)

64.08984995278189


**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [48]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])

Coefficient
Alaska             2.182355e+13
Arizona            1.784608e+13
California         5.612804e+13
Colorado           5.735584e+13
Connecticut        2.808780e+13
Idaho              4.304227e+13
Illinois           2.516111e+13
Indiana            1.784608e+13
Iowa               2.182355e+13
Maine              3.745111e+13
Maryland           1.263831e+13
Massachusetts      4.127458e+13
Michigan           6.509884e+13
Minnesota          4.634454e+13
Missouri           1.784608e+13
Montana            4.304227e+13
Nevada             2.516111e+13
New Hampshire      4.938736e+13
New Jersey         1.784608e+13
New Mexico         3.745111e+13
New York           6.898047e+13
North Carolina     3.072127e+13
Ohio               2.808780e+13
Oregon             3.941540e+13
Pennsylvania       5.356091e+13
Rhode Island       1.263831e+13
South Dakota       1.784608e+13
Tennessee          1.263831e+13
Utah               4.472932e+13
Vermont            4.789519e+13
Virginia           2.516111e+13
Washington         3.941540e+13
West Virginia      2.516111e+13
Wisconsin          4.938736e+13
Wyoming            3.536421e+13
Unnamed: 0         6.468750e+00
summit_elev        1.925781e+00
vertical_drop      1.167969e+00
trams              4.594020e+12
fastEight          4.516371e+11
fastSixes          5.346684e+12
fastQuads          1.803567e+13
quad               1.076617e+13
triple             1.328398e+13
double             1.489121e+13
surface            1.689807e+13
total_chairs       4.757469e+13
Runs               1.274414e+00
TerrainParks       2.636719e-02
LongestRun_mi      3.828125e-01
SkiableTerrain_ac  5.429688e-01
Snow Making_ac     2.070312e-01
daysOpenLastYear   5.244141e-01
yearsOpen          4.023438e-01
averageSnowfall    8.203125e-02
AdultWeekday       1.933203e+01
projectedDaysOpen  1.425781e-01
NightSkiing_ac     7.890625e-01
clusters           9.492188e-01

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [49]:
df2 = df.drop('state', axis=1)

In [50]:
df2.shape
df2.describe().T

count         mean          std    min      25%  \
Unnamed: 0         330.0   164.500000    95.407023    0.0    82.25   
summit_elev        330.0  4591.818182  3735.535934  315.0  1403.75   
vertical_drop      330.0  1215.427273   947.864557   60.0   461.25   
trams              330.0     0.172727     0.559946    0.0     0.00   
fastEight          330.0     0.003030     0.055048    0.0     0.00   
fastSixes          330.0     0.184848     0.651685    0.0     0.00   
fastQuads          330.0     1.018182     2.198294    0.0     0.00   
quad               330.0     0.933333     1.312245    0.0     0.00   
triple             330.0     1.500000     1.619130    0.0     0.00   
double             330.0     1.833333     1.815028    0.0     1.00   
surface            330.0     2.621212     2.059636    0.0     1.00   
total_chairs       330.0     8.266667     5.798683    0.0     5.00   
Runs               330.0    48.214724    46.081366    3.0    19.00   
TerrainParks       330.0     2.820789     1.845918    1.0     2.00   
LongestRun_mi      330.0     1.433231     1.147352    0.0     0.60   
SkiableTerrain_ac  330.0   739.801223  1807.868079    8.0    85.00   
Snow Making_ac     330.0   174.873239   242.378848    2.0    55.00   
daysOpenLastYear   330.0   115.103943    32.231212    3.0   100.00   
yearsOpen          330.0    63.656535   109.263494    6.0    50.00   
averageSnowfall    330.0   185.316456   133.424094   18.0    71.25   
AdultWeekday       330.0    57.916957    23.898798   15.0    43.00   
AdultWeekend       330.0    64.166810    22.571324   17.0    49.00   
projectedDaysOpen  330.0   120.053004    28.742976   30.0   101.25   
NightSkiing_ac     330.0    56.890909    93.464882    0.0     0.00   
clusters           330.0     0.651515     0.483541    0.0     0.00   

                           50%          75%      max  
Unnamed: 0          164.500000   246.750000    329.0  
summit_elev        3127.500000  7806.000000  13487.0  
vertical_drop       964.500000  1800.000000   4425.0  
trams                 0.000000     0.000000      4.0  
fastEight             0.000000     0.000000      1.0  
fastSixes             0.000000     0.000000      6.0  
fastQuads             0.000000     1.000000     15.0  
quad                  0.000000     1.000000      8.0  
triple                1.000000     2.000000      8.0  
double                1.000000     3.000000     14.0  
surface               2.000000     3.000000     15.0  
total_chairs          7.000000    10.000000     41.0  
Runs                 34.000000    60.000000    341.0  
TerrainParks          2.820789     3.000000     14.0  
LongestRun_mi         1.000000     2.000000      6.0  
SkiableTerrain_ac   200.000000   739.801223  26819.0  
Snow Making_ac      132.500000   192.000000   3379.0  
daysOpenLastYear    115.103943   130.000000    305.0  
yearsOpen            58.000000    68.750000   2019.0  
averageSnowfall     158.000000   278.750000    669.0  
AdultWeekday         57.916957    67.000000    179.0  
AdultWeekend         64.166810    72.000000    179.0  
projectedDaysOpen   120.053004   136.000000    305.0  
NightSkiing_ac       27.000000    83.750000    650.0  
clusters              1.000000     1.000000      2.0

In [51]:
# Declare an explanatory variable, called X2,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X2 = df2.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y2 = df2.AdultWeekend 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X2)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X2_scaled=scaler.transform(X2) 

In [52]:
# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y2 = y2.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X2_train, X2_test, y2_train, y2_test = train_test_split(X2_scaled, y2, test_size=0.25, random_state=1)

In [53]:
lm2 = linear_model.LinearRegression()
model2 = lm2.fit(X2_train,y2_train)

In [54]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y2_pred = model2.predict(X2_test)

In [55]:
print(explained_variance_score(y2_test, y2_pred))

0.9311231709683413


In [56]:
print(mean_absolute_error(y2_test, y2_pred))

5.315966330606209


In [57]:
print(lm2.intercept_)

64.04905616806636


In [58]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm2.coef_), X2.columns, columns=['Coefficient'])

Coefficient
Unnamed: 0            0.592733
summit_elev           4.348826
vertical_drop         0.954508
trams                 0.063597
fastEight             0.190794
fastSixes             0.465249
fastQuads             1.169446
quad                  1.029905
triple                1.489882
double                0.227721
surface               1.336423
total_chairs          0.736969
Runs                  1.074032
TerrainParks          0.134543
LongestRun_mi         0.321581
SkiableTerrain_ac     1.787890
Snow Making_ac        0.150218
daysOpenLastYear      0.894095
yearsOpen             0.295443
averageSnowfall       1.016455
AdultWeekday         19.766083
projectedDaysOpen     0.338348
NightSkiing_ac        0.785930
clusters              5.289804

### Model 3

In [59]:
df3 = df.drop(['state', 'summit_elev'], axis=1)

In [60]:
df3.shape
df3.describe().T

count         mean          std   min     25%         50%  \
Unnamed: 0         330.0   164.500000    95.407023   0.0   82.25  164.500000   
vertical_drop      330.0  1215.427273   947.864557  60.0  461.25  964.500000   
trams              330.0     0.172727     0.559946   0.0    0.00    0.000000   
fastEight          330.0     0.003030     0.055048   0.0    0.00    0.000000   
fastSixes          330.0     0.184848     0.651685   0.0    0.00    0.000000   
fastQuads          330.0     1.018182     2.198294   0.0    0.00    0.000000   
quad               330.0     0.933333     1.312245   0.0    0.00    0.000000   
triple             330.0     1.500000     1.619130   0.0    0.00    1.000000   
double             330.0     1.833333     1.815028   0.0    1.00    1.000000   
surface            330.0     2.621212     2.059636   0.0    1.00    2.000000   
total_chairs       330.0     8.266667     5.798683   0.0    5.00    7.000000   
Runs               330.0    48.214724    46.081366   3.0   19.00   34.000000   
TerrainParks       330.0     2.820789     1.845918   1.0    2.00    2.820789   
LongestRun_mi      330.0     1.433231     1.147352   0.0    0.60    1.000000   
SkiableTerrain_ac  330.0   739.801223  1807.868079   8.0   85.00  200.000000   
Snow Making_ac     330.0   174.873239   242.378848   2.0   55.00  132.500000   
daysOpenLastYear   330.0   115.103943    32.231212   3.0  100.00  115.103943   
yearsOpen          330.0    63.656535   109.263494   6.0   50.00   58.000000   
averageSnowfall    330.0   185.316456   133.424094  18.0   71.25  158.000000   
AdultWeekday       330.0    57.916957    23.898798  15.0   43.00   57.916957   
AdultWeekend       330.0    64.166810    22.571324  17.0   49.00   64.166810   
projectedDaysOpen  330.0   120.053004    28.742976  30.0  101.25  120.053004   
NightSkiing_ac     330.0    56.890909    93.464882   0.0    0.00   27.000000   
clusters           330.0     0.651515     0.483541   0.0    0.00    1.000000   

                           75%      max  
Unnamed: 0          246.750000    329.0  
vertical_drop      1800.000000   4425.0  
trams                 0.000000      4.0  
fastEight             0.000000      1.0  
fastSixes             0.000000      6.0  
fastQuads             1.000000     15.0  
quad                  1.000000      8.0  
triple                2.000000      8.0  
double                3.000000     14.0  
surface               3.000000     15.0  
total_chairs         10.000000     41.0  
Runs                 60.000000    341.0  
TerrainParks          3.000000     14.0  
LongestRun_mi         2.000000      6.0  
SkiableTerrain_ac   739.801223  26819.0  
Snow Making_ac      192.000000   3379.0  
daysOpenLastYear    130.000000    305.0  
yearsOpen            68.750000   2019.0  
averageSnowfall     278.750000    669.0  
AdultWeekday         67.000000    179.0  
AdultWeekend         72.000000    179.0  
projectedDaysOpen   136.000000    305.0  
NightSkiing_ac       83.750000    650.0  
clusters              1.000000      2.0

In [61]:
# Declare an explanatory variable, called X2,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X3 = df3.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y3 = df3.AdultWeekend 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X3)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X3_scaled=scaler.transform(X3) 

In [62]:
# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y3 = y3.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X3_train, X3_test, y3_train, y3_test = train_test_split(X3_scaled, y3, test_size=0.25, random_state=1)

In [63]:
lm3 = linear_model.LinearRegression()
model3 = lm3.fit(X3_train,y3_train)

In [64]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y3_pred = model3.predict(X3_test)

In [65]:
print(explained_variance_score(y3_test, y3_pred))

0.9335698986130937


In [66]:
print(mean_absolute_error(y3_test, y3_pred))

5.221526943535906


In [67]:
print(lm3.intercept_)

64.04892455911474


In [68]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm3.coef_), X3.columns, columns=['Coefficient'])

Coefficient
Unnamed: 0            0.618138
vertical_drop         1.893158
trams                 0.295871
fastEight             0.011682
fastSixes             0.455371
fastQuads             0.797188
quad                  1.122427
triple                1.491350
double                0.252935
surface               1.134756
total_chairs          0.671475
Runs                  0.908958
TerrainParks          0.359699
LongestRun_mi         0.249915
SkiableTerrain_ac     0.832256
Snow Making_ac        0.111185
daysOpenLastYear      1.326494
yearsOpen             0.255431
averageSnowfall       0.875661
AdultWeekday         20.028212
projectedDaysOpen     0.605334
NightSkiing_ac        0.382499
clusters              2.419525

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.944 | 4.726  |-|
| Model 2. | 0.931| 5.316 |'state'|
| Model 3. | 0.934 | 5.222 |'state','summit_elev','base_elev'|

Model Selection:

In [ ]:
# Model 3 - It removes inherent features that the resort cannot control. 